File: backend/app/schemas/story.py
from typing import List, Optional, Dict
from pydantic import BaseModel, ConfigDict
from datetime import datetime

class StoryOptionsSchema(BaseModel):
    text: str
    node_id: Optional[int] = None


#Parent class
class StoryNodeBase(BaseModel):
    content: str
    is_ending: bool = False
    is_winning_ending: bool = False

class CompleteStoryNodeResponse(StoryNodeBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    options: List[StoryOptionsSchema] = []


class StoryBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    title: str
    session_id: Optional[str] = None


class CreateStoryRequest(BaseModel):
    theme: str


class CompleteStoryResponse(StoryBase):

    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    root_node: CompleteStoryNodeResponse
    all_nodes: Dict[int, CompleteStoryNodeResponse]

File: frontend/src/types/story.ts
export interface StoryOption {
    text: string
    node_id: string
}

export interface StoryNode {
    id: number
    content: string
    is_ending: boolean
    is_winning_ending: boolean
    options: StoryOption[]
}

export interface Story {
    id: number
    title: string
    session_id: string
    created_at: string
    root_node: StoryNode
    all_nodes: Record<string, StoryNode>
}

export interface CreateStoryRequest {
    theme: string
}

export interface StoryJob {
    job_id: string
    status: "Pending" | "processing" | "completed" | "failed"
    created_at: string
    story_id: number | null
    completed_at: string | null
    error: string | null
}
